### PR TITLE
encode the package version in the shared object

### LINKED ISSUE #4754: Encode package version as a shared object symbol
One of the more common issues we see with users is an `AttributeError` where one of our `Cryptograpy_HAS_*` can't be found. This is pretty much always an issue where Python has loaded a newer version of cryptography's Python code, but for some reason loads an older shared object (which does not have the symbol the Python code is looking for).

This is a bad error message for a serious environment problem, and it doesn't always happen immediately. We should encode the version of cryptography in the shared object we build and then raise an `ImportError` on import if the shared object version and cryptography's version do not match.

refs https://github.com/certbot/certbot/issues/5651

### PR DESCRIPTION
**Pull request by reaperhulk**
fixes #4754

### CODE REVIEW

File: src/cryptography/hazmat/bindings/openssl/binding.py [hunk 1]
```diff
@@ -164,6 +165,29 @@ def _verify_openssl_version(lib):
...
+    so_package_version = Binding.ffi.string(
+        Binding.lib.CRYPTOGRAPHY_PACKAGE_VERSION
+    ).decode("ascii")
+    if version != so_package_version:
+        raise ImportError(
+            "The version of cryptography does not match with the loaded "
+            "shared object. This can happen if you have multiple copies of "
+            "cryptography installed in your Python path. Please try creating "
+            "a new virtual environment to resolve this issue. "
+            "Loaded python version: {0}, shared object version: {1}".format(
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/binding.py]: Now that we've dropped py2.6 you can just use `{}` you don't need to number them

File: src/_cffi_src/openssl/cryptography.py [hunk 2]
```diff
@@ -4,6 +4,19 @@
...
 
+import os
+
+
+# Load the cryptography __about__ to get the current package version
+base_src = os.path.dirname(
+    os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
+)
+about = os.path.join(base_src, "cryptography", "__about__.py")
+__about__ = {}
```
Comment (alex) [inline src/_cffi_src/openssl/cryptography.py]: There's no reason for this to have a weird `__about__` name

File: src/_cffi_src/openssl/cryptography.py [hunk 3]
```diff
@@ -62,7 +75,8 @@
     (OPENSSL_VERSION_NUMBER < 0x101000af || CRYPTOGRAPHY_IS_LIBRESSL)
 #define CRYPTOGRAPHY_OPENSSL_LESS_THAN_111 \
     (OPENSSL_VERSION_NUMBER < 0x10101000 || CRYPTOGRAPHY_IS_LIBRESSL)
-"""
+#define CRYPTOGRAPHY_PACKAGE_VERSION "{0}"
```
Comment (alex) [inline src/_cffi_src/openssl/cryptography.py]: same

File: src/_cffi_src/openssl/cryptography.py [hunk 4]
```diff
@@ -62,7 +74,8 @@
     (OPENSSL_VERSION_NUMBER < 0x101000af || CRYPTOGRAPHY_IS_LIBRESSL)
 #define CRYPTOGRAPHY_OPENSSL_LESS_THAN_111 \
     (OPENSSL_VERSION_NUMBER < 0x10101000 || CRYPTOGRAPHY_IS_LIBRESSL)
-"""
+#define CRYPTOGRAPHY_PACKAGE_VERSION "{}"
+""".format(about["__version__"])
```
Comment (alex) [inline src/_cffi_src/openssl/cryptography.py]: Would it make more sense to do this in _all_ `build_ffi_for_binding`?

File: src/cryptography/hazmat/bindings/openssl/binding.py [hunk 5]
```diff
@@ -164,6 +165,29 @@ def _verify_openssl_version(lib):
...
+    # are installed and available in the python path. To avoid errors cropping
+    # up later this code checks that the currently imported package and the
+    # shared object that were loaded have the same version and raise an
+    # ImportError if they do not
+    so_package_version = Binding.ffi.string(
+        Binding.lib.CRYPTOGRAPHY_PACKAGE_VERSION
+    ).decode("ascii")
+    if version != so_package_version:
+        raise ImportError(
+            "The version of cryptography does not match with the loaded "
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/binding.py]:
```suggestion
            "The version of cryptography does not match the loaded "
```

File: src/cryptography/hazmat/bindings/openssl/binding.py [hunk 6]
```diff
@@ -172,3 +196,4 @@ def _verify_openssl_version(lib):
 Binding.init_static_locks()
 
 _verify_openssl_version(Binding.lib)
+_verify_package_version(cryptography.__version__)
```
Comment (alex) [inline src/cryptography/hazmat/bindings/openssl/binding.py]: Verify cryptography version first I think? Before the locks even.

**Comment by bmw** [comment]
I think the `ImportError` being raised here will help us a lot when people hit this issue. Thank you for adding it!

**Pull request merged**